### PR TITLE
Refactor ldap strategy

### PR DESCRIPTION
**Pull request by bgeuken**

### CODE REVIEW

File: src/api/app/models/user_ldap_strategy.rb [hunk 1]
```diff
@@ -591,4 +582,18 @@ def self.try_ldap_con(server, user_name, password)
     Rails.logger.debug("Bound as #{user_name}")
     conn
   end
+
+  private
+
+  def check_groups(relationships, method_name)
```
Comment (evanrolfe) [inline src/api/app/models/user_ldap_strategy.rb]: Would be nice to have a more descriptive name, i.e. `relationship_groups_contains_user?`

File: src/api/spec/models/user_ldap_strategy_spec.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  let(:user) { create(:user, login: 'eisendieter') }
+  let(:confirmed_user) { create(:confirmed_user, login: 'confirmed_user') }
+  let(:input) { { 'Event::RequestCreate' => { source_maintainer: '1' } } }
+  let(:project_with_package) { create(:project_with_package, name: 'project_b') }
+
+  describe '.dn2user_principal_name' do
+    context 'when no user id is provided' do
+      it 'returns an empty string' do
+        expect(UserLdapStrategy.dn2user_principal_name('cn=jsmith,ou=Promotions,dc=noam,dc=com')).to eq('')
+        expect(UserLdapStrategy.dn2user_principal_name(['cn=jsmith,ou=Promotions,dc=noam,dc=com'])).to eq('')
```
Comment (evanrolfe) [inline src/api/spec/models/user_ldap_strategy_spec.rb]: Would be nice to use a variable here like this:
```ruby
let(:dn_string) { 'cn=jsmith,ou=Promotions,dc=noam,dc=com' }

it 'returns an empty string' do
  expect(UserLdapStrategy.dn2user_principal_name(dn_string)).to eq('')
  expect(UserLdapStrategy.dn2user_principal_name([dn_string])).to eq('')
end
```

Comment (bgeuken) [inline src/api/spec/models/user_ldap_strategy_spec.rb]: Good idea. Looks much better now.

File: src/api/app/models/user_ldap_strategy.rb [hunk 3]
```diff
@@ -591,4 +582,18 @@ def self.try_ldap_con(server, user_name, password)
...
+  private
+
+  def relationship_groups_contains_user?(relationships, method_name)
+    relationships.each do |relationship|
+      return false if relationship.group.nil?
+      # check whether current user is in this group
+      return true if user_in_group_ldap?(login, relationship.group)
+    end
+
+    Rails.logger.info "Failed with #{method_name}"
```
Comment (mdeniz) [inline src/api/app/models/user_ldap_strategy.rb]: [nitpick] Why not using `Kernel#caller` for calculating the method_name ?

```ruby
method_name = caller[0][/`.*'/][1..-2]
```

Comment (bgeuken) [inline src/api/app/models/user_ldap_strategy.rb]: To me the current code is more straight forward / easier to understand. I would prefer to keep it this way, or to remove the method name reference. We probably don't need it.

But we can still change this later.

**Pull request merged**